### DISCUSSION TITLE
How to update reactive dictionary?

### DISCUSSION THREAD

**Discussion by Ljupch0** [widgetti/solara#227] [General]
I have the following reactive dict:
```
flex_attribute_values_selected = solara.reactive({})

@solara.component
def flex_attributes_value_selection(df, flex_attribute):
  solara.SelectMultiple(
    label=flex_attribute,
    values = flex_attribute_values_selected,
    all_values = list(df[flex_attribute].unique())
    )
```

`flex_attributes_value_selection` is a select multiple, I want to push the values of this multiple selection to the reactive dict `flex_attribute_values_selected` under a new key, which would be the `flex_attribute` provided to the component.

I tried many things but going in circles.

**Comment by maartenbreddels** [comment]
Hi Ljupcho,

it is not clear to me what the question is.


> `flex_attributes_value_selection` is a select multiple

What do you mean by this. `flex_attributes_value_selection` to me looks like a reactive value that wraps a dictionary.


> values = flex_attribute_values_selected,

If you use vscode with the Python extension (which uses pylance), or if you use mypy, you will see this is not correctly typed. Values should be a List or Reactive[List], not a dict. 

I'm guessing here, that you want `flex_attribute_values_selected[flex_attribute]` to be set to the selected items?

You didn't give me a full reproducible example, so I'm guessing you want sth like this:

```python
import pandas as pd

import solara

df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
flex_attribute_values_selected = solara.reactive({})


@solara.component
def flex_attributes_value_selection(df, flex_attribute):
    print("flex_attributes_value_selection", flex_attribute_values_selected.value)

    def update(values):
        print("update", values, flex_attribute)
        new_dict = {**flex_attribute_values_selected.value, flex_attribute: values}
        flex_attribute_values_selected.value = new_dict

    solara.SelectMultiple(
        label=flex_attribute,
        values=flex_attribute_values_selected.value.get(flex_attribute, []),
        all_values=list(df[flex_attribute].unique()),
        on_value=update,
    )


@solara.component
def Page():
    flex_attribute = solara.use_reactive("a")
    print("Page", flex_attribute_values_selected.value)
    solara.Select("flex attribute", ["a", "b"], value=flex_attribute)

    flex_attributes_value_selection(df, flex_attribute.value)
    md = "## Values: \n```\n" + repr(flex_attribute_values_selected.value) + "\n```"
    print(md)
    solara.Markdown(md)

```

**Comment by Ljupch0** [comment]
Amazing, this is exactly what I needed🙏

The `on_value` update logic is what I was missing. I will look into it in depth.

Sorry for bad and not repoducible example Maarten, better next time.

**Comment by Ljupch0** [comment]
On second pass, I now realize that I'm trying to reimplement the functionality of CrossFilter. Basically select columns and apply filters by selecting values. The reactive dictionary was supposed to hold the column name and selected values, and use that to filter the dataframe. Great to see batteries included for this.

**Reply by maartenbreddels** [reply]
Good to hear. Still, a good practice to try to build it yourself.

**Reply by Ljupch0** [reply]
I am not clear on how to access the filtered dataframe. With `solara.CrossFilterDataFrame(filtered_dataframe)` I can see the dataframe and the correct filters are being applied. I want to take `filtered_dataframe` and use it to chart in matplotlib or other, and have the visual update as filters are added/removed. Is this not how it's supposed to work? Currently I have it with matplotlib and the visual is not reacting to filter changes, I have `dependency = filtered_dataframe`. I will come back with reproducible, just to get your take.

**Reply by maartenbreddels** [reply]
Yes, you can copy the same pattern in your own component. You receive the same filter as the the component you refer to. Did you try this?
